cmd/oras/blob/delete.go

```diff
@@ -0,0 +1,88 @@
+/*
+Copyright The ORAS Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package blob
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/spf13/cobra"
+	"oras.land/oras-go/v2/errdef"
+	"oras.land/oras/cmd/oras/internal/option"
+)
+
+type deleteBlobOptions struct {
+	option.Common
+	option.Remote
+
+	targetRef string
+}
+
+func deleteCmd() *cobra.Command {
+	var opts deleteBlobOptions
+	cmd := &cobra.Command{
+		Use:   "delete <name@digest> [flags]",
+		Short: "[Preview] Delete a blob from a remote registry",
+		Long: `[Preview] Delete a blob from a remote registry
+** This command is in preview and under development. **
+
+Example - Delete blob:
+  oras blob delete localhost:5000/hello@sha256:9a201d228ebd966211f7d1131be19f152be428bd373a92071c71d8deaf83b3e5
+
+Example - Delete blob from the insecure registry:
+  oras blob delete localhost:5000/hello@sha256:9a201d228ebd966211f7d1131be19f152be428bd373a92071c71d8deaf83b3e5 --insecure
+`,
+		Args: cobra.ExactArgs(1),
+		PreRunE: func(cmd *cobra.Command, args []string) error {
+			return opts.ReadPassword()
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			opts.targetRef = args[0]
+			return deleteBlob(opts)
+		},
+	}
+
+	option.ApplyFlags(&opts, cmd.Flags())
+	return cmd
+}
+
+func deleteBlob(opts deleteBlobOptions) (err error) {
+	ctx, _ := opts.SetLoggerLevel()
+	repo, err := opts.NewRepository(opts.targetRef, opts.Common)
+	if err != nil {
+		return err
+	}
+
+	if repo.Reference.Reference == "" || !strings.Contains(opts.targetRef, "@") {
```

Comment:
There is a better way to check whether it is a digest or not.

See https://pkg.go.dev/oras.land/oras-go/v2@v2.0.0-rc.2.0.20220901121748-295b1c32b079/registry#Reference.Digest

Reply:
`Reference.Digest()` checks whether the reference is a validated digest object. However, if the target ref is like `localhost:5000/0819hello:sha256:ca85dbbdd91b589ebecf4a67b015eab7a629e0d619dcb6aff48d7d856770dd1b`, the error will not be thrown because the reference string is still `sha256:ca85dbbdd91b589ebecf4a67b015eab7a629e0d619dcb6aff48d7d856770dd1b`. Do we need to consider this case?

Reply:
We don't need to consider this case as it will be addressed by https://github.com/oras-project/oras-go/issues/303